File: 66.cpp
class Solution {
public:
    vector<int> plusOne(vector<int>& digits) {
        int n = digits.size();
        int carry = 1;
        for (int i = n - 1; i >= 0; i--) {
            if (carry != 0) {
                if (digits[i] == 9) {
                    digits[i] = 0;
                }
                else {
                    digits[i] += 1;
                    carry = 0;
                }
            }
            else {
                break;
            }
        }
        if (carry != 0) {
            // The number was all 9s
            digits.resize(n+1);
            digits[n] = 0;
            digits[0] = 1;
        }
        return digits;        
    }
};

File: 66.cs
public class Solution {
    public int[] PlusOne(int[] digits) {
        int n = digits.Length;
        for (int i = n - 1; i >= 0; i--)  {
            if (digits[i] != 9) {
                digits[i] += 1;
                break;
            }
            else {
                digits[i] = 0;
            }
        }
        if (digits[0] == 0) {
            int[] new_num = new int[n + 1];
            new_num[0] = 1;
            for (int j = 1; j <= n; j++) {
                new_num[j] = 0;
            }
            return new_num;
        }
        return digits;      
    }
}
